transport/happyeyeballs_test.go

```diff
@@ -0,0 +1,283 @@
+// Copyright 2024 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transport
+
+import (
+	"context"
+	"errors"
+	"net/netip"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+type colletcStreamDialer struct {
+	Dialer StreamDialer
+	Addrs  []string
+}
+
+func (d *colletcStreamDialer) DialStream(ctx context.Context, addr string) (StreamConn, error) {
+	d.Addrs = append(d.Addrs, addr)
+	return d.Dialer.DialStream(ctx, addr)
+}
+
+var nilDialer = FuncStreamDialer(func(ctx context.Context, addr string) (StreamConn, error) {
+	return nil, nil
+})
+
+func newErrorStreamDialer(err error) StreamDialer {
+	return FuncStreamDialer(func(ctx context.Context, addr string) (StreamConn, error) {
+		return nil, err
+	})
+}
+
+func TestHappyEyeballsStreamDialer_DialStream(t *testing.T) {
+	t.Run("Works with IPv4 hosts", func(t *testing.T) {
```

Comment:
What's the benefit of using `t.Run` instead of defining a test function `func TestIPv4HostShouldWork() {}`?